COLOS-21: Add executioner

iboshkov opened the pull request:
## Issue ID

https://plane.42.mk/colosseum/projects/feca94dd-2b65-430b-b917-9299b2f82c60/issues/952a7950-caba-4fdf-b118-c5572085a013

## What has been done?

 - [x] Added initial implementation for executioner 

## Sceenshots/Videos

<!--Add screenshots or videos of the current task-->

iboshkov committed:
Add rough implementation for executioner

iboshkov committed:
Make max prefetched messages configurable

iboshkov committed:
wow.gif

iboshkov committed:
Fix senatus

iboshkov committed:
Adding processing logic to senatus, start script, fixing gcc/g++

iboshkov committed:
Fix executioner image references

iboshkov committed:
Add Python language support

Aleksandar1932 commented on executioner/languages/docker.ts:
```diff
@@ -0,0 +1,170 @@
+        `${compileStderrPath}:/exc/${compileStderrFilename}`,
+        "-i",
+        `${IMAGE_BASE}${lang}`
+      ];
```
Please add resource limits, to prevent infinite loops and memory leaks https://docs.docker.com/config/containers/resource_constraints/

iboshkov replied:
Added in each entrypoint.sh

vergjor committed:
Merge branch 'development' into feature/executioner

vergjor committed:
Update build.yml to trigger on pull requests with changes within the listed paths

vergjor committed:
Filter the push events as well based on the files changes

iboshkov committed:
Keep snapshot of code

vergjor committed:
Improve the CI yaml files

iboshkov committed:
Make image base url dynamic

iboshkov committed:
Push for dev branch

iboshkov committed:
Merge remote-tracking branch 'origin/feature/executioner' into feature/executioner

iboshkov committed:
Fix build context location

iboshkov committed:
Enable push globally

iboshkov committed:
Add push permissions

iboshkov committed:
Add global permissions

gjorgi-krenkov commented on executioner/runtimes/languages/gcc/entrypoint.sh:
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+# Compile the source code
+ERROR=$(g++ -O2 -static src/main.cpp -o ./main 2>&1)
```
We also need to be aware of whether we are compiling C++23 or C++17 code, for example, probably the same for other languages.

iboshkov replied:
Not important ATM

gjorgi-krenkov commented on senatus/src/api/submission/routes/0-submit.ts:
```diff
@@ -0,0 +1,13 @@
+export default  {
+  routes: [
+    {
+      method: 'GET',
```
Why is this method GET?

iboshkov replied:
Changed to POST

gjorgi-krenkov commented on senatus/src/index.ts:
```diff
@@ -14,5 +17,33 @@ export default {
+            executionTime: result.time,
+            processed: true,
+            processedAt: new Date(),
+            publishedAt: new Date(),
```
Why are these two identical?

iboshkov replied:
PublishedAt is the Strapi way of publishing an entity, processed at is our own indicator for when the execution actually finished.

gjorgi-krenkov commented on senatus/src/index.ts:
```diff
@@ -14,5 +17,33 @@ export default {
+    } = await connect(strapi.config.get("server.app.rabbitUrl"));
+    console.log("Connected to RabbitMQ!");
+    await subscribe("results", async (msg) => {
+      const parsed = JSON.parse(msg);
```
Can we validate the message at this point?

gjorgi-krenkov commented on senatus/src/api/submission/controllers/submission.ts:
```diff
@@ -2,6 +2,96 @@
+    try {
+      ctx.body = ctx.request.query;
+
+      const id = parseInt(ctx.request.query.id as string);
```
Isn't it more convenient to make this method to also store a submission if it's not provided?

iboshkov replied:
It is, but I would leave it out for now.

iboshkov committed:
Adding execution timeout, Java, C# environments

iboshkov committed:
Add build for new execution environments

iboshkov committed:
Change method of submit to POST

iboshkov committed:
Add passed update to senatus

iboshkov commented:
Added:

- Execution timeout using the `timeout` GNU CoreUtils program
- Added Java
- Added C#
- Added support for piping compilation errors into stderr

iboshkov committed:
Adding missing path checks

iboshkov commented:
- Added C#
- Added Java
- Split AMQP connection into pub/sub channels
- Piping compilation errors into stderr on compiled languages

iboshkov committed:
Adding rust

iboshkov committed:
Fix mindepth vs maxdepth usage. Adding golang. Adding TypeScript to NodeJS

iboshkov committed:
Adding golang build

iboshkov committed:
Fix C#
